### PR TITLE
some speedup features

### PR DESCRIPTION
**Pull request by melsman**
Several speedup features:

- [x] flattening of scoring using expand
- [x] simpler determination of whether a point is in a triangle
- [x] generate smaller objects when more objects have been drawn (big objects will likely fail)
- [ ] initiate work on parallelising the generation and drawing of objects (currently doesn't work, primarily due to a compiler limitation)

### CODE REVIEW

**Comment by melsman** [comment]
image

image

(After a night's work - for my laptop, that is... ;)

File: shapegen.fut [hunk 1]
```diff
@@ -29,34 +30,83 @@ module mk_full_shape (o: shape) = {
...
+
+      let best_try (s0, sq0) (s1, sq1) = if s0 > s1 then (s0, sq0) else (s1, sq1)
+      let (score_best, sq_best) = reduce_comm best_try (-f32.inf, o.empty) tries
+      let rng = rnge.join_rng rngs
+      in (rng, (j,i,score_best, sq_best))
+
+    --let G = 1
+    --let (hG,wG) = (h / G, w / G)
+    --let (rng, res:[][](i32,i32,f32,o.t)) = gridify G rng (on_grid (hG,wG))
+    --let (image_approx', improvement) = render2 image_approx (hG,wG) (flatten res)
```
Comment (nqpz) [inline shapegen.fut]: It seems you left in some experiment code here.  Please clean it up.

**Review by nqpz — CHANGES_REQUESTED** [review]
Neat!  Certainly seems to generate better pictures now.

**Comment by nqpz** [comment]
Thanks again!  I really hope we can get some kind of parallelised object drawing going.

**Comment by nqpz** [comment]
Thanks again!  I really hope we can get some kind of parallelised object drawing going.

**Comment by melsman** [comment]
Hi there,

I'm just about getting it to work! ;) But let me know if you're also doing
it...

Just wait with merging the PR...

Best,

Martin

tir. 25. feb. 2020 kl. 18.27 skrev Niels G. W. Serup <
notifications@github.com>:

**Comment by nqpz** [comment]
Cool! I'm not working on it, so I'm looking forward to it.

**Pull request merged**